File: go/plugins/gcpsecrets/gcpsecrets.go
// The gcpsecrets plugin: GCP Secret Manager. Needs HTTPS. A port of
// typescript/plugins/gcpsecrets.ts.
package gcpsecrets

import (
	"encoding/base64"
	"net/http"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/voxgig/sekreto/go/plugins/httpjson"
	"github.com/voxgig/sekreto/go/sekreto"
)

// Provider reads GCP Secret Manager.
//
// api.token reads secret api_token (dots flattened to `_`; Secret Manager
// ids have no hierarchy and reject dots), latest version. The token comes
// from config, then GOOGLE_OAUTH_ACCESS_TOKEN, then the GCE/GKE metadata
// server - so on Google's own platform no credential configuration is
// needed at all.
//
// The metadata call itself is plain http to a link-local host by platform
// design; no credential rides on it, so checkaddr guards the Secret
// Manager address instead.
type Provider struct {
	Project      string
	Token        string
	Addr         string
	MetadataAddr string
	// A configured token is kept forever; a metadata-server token carries
	// expires_in and is renewed shortly before it runs out.
	livetoken string
	havetoken bool
	renewat   time.Time
	mu        sync.Mutex
}

func (provider *Provider) metadataaddr() string {
	if "" != provider.MetadataAddr {
		return provider.MetadataAddr
	}
	if host := os.Getenv("GCE_METADATA_HOST"); "" != host {
		return "http://" + host
	}
	return "http://metadata.google.internal"
}

func (provider *Provider) login() (string, error) {
	configured := provider.Token
	if "" == configured {
		configured = os.Getenv("GOOGLE_OAUTH_ACCESS_TOKEN")
	}
	if "" != configured {
		return configured, nil
	}

	target := strings.TrimSuffix(provider.metadataaddr(), "/") +
		"/computeMetadata/v1/instance/service-accounts/default/token"

	status, body, err := httpjson.Get(target, map[string]string{"Metadata-Flavor": "Google"})
	if nil != err {
		return "", err
	}

	got := httpjson.DigText(body, "access_token")
	if http.StatusOK != status || "" == got {
		return "", sekreto.Fail("sekreto: gcp: no token and metadata server did not answer")
	}

	provider.renewat = httpjson.Expiry(httpjson.Dig(body, "expires_in"))

	return got, nil
}

func (provider *Provider) Lookup(name string) (string, bool, error) {
	// Serialize per-provider access: Lookup reads and refreshes the cached
	// login token, and a secrets client may resolve from several goroutines.
	provider.mu.Lock()
	defer provider.mu.Unlock()

	if "" == provider.Project {
		return "", false, sekreto.Fail("sekreto: gcp: no project")
	}

	addr := provider.Addr
	if "" == addr {
		addr = "https://secretmanager.googleapis.com"
	}
	if err := sekreto.CheckAddr(addr); nil != err {
		return "", false, err
	}

	if !provider.havetoken || httpjson.Due(provider.renewat) {
		token, err := provider.login()
		if nil != err {
			return "", false, err
		}
		provider.livetoken = token
		provider.havetoken = true
	}

	flat, err := sekreto.FlatName(name, "_")
	if nil != err {
		return "", false, err
	}

	target := strings.TrimSuffix(addr, "/") + "/v1/projects/" + provider.Project +
		"/secrets/" + flat + "/versions/latest:access"

	status, body, err := httpjson.Get(target,
		map[string]string{"authorization": "Bearer " + provider.livetoken})
	if nil != err {
		return "", false, err
	}

	if http.StatusNotFound == status {
		return "", false, nil
	}

	if http.StatusOK != status {
		return "", false, sekreto.Fail("sekreto: gcp error: " + strconv.Itoa(status) + ": " + target)
	}

	data, is := httpjson.Dig(body, "payload", "data").(string)
	if !is {
		return "", false, nil
	}

	// See the aws provider: an undecodable payload is an error, not a miss.
	decoded, err := base64.StdEncoding.DecodeString(data)
	if nil != err {
		return "", false, sekreto.Fail("sekreto: gcp: undecodable secret")
	}
	return string(decoded), true, nil
}

func (provider *Provider) Describe() string {
	return "gcpsecrets:" + provider.Project
}

// Plugin is the `gcpsecrets` provider kind, as a voxgig/plugin definition.
var Plugin = sekreto.ProviderPlugin("gcpsecrets", func(spec *sekreto.ProviderSpec) (sekreto.Provider, error) {
	return &Provider{
		Project:      spec.Project,
		Token:        spec.Token,
		Addr:         spec.Addr,
		MetadataAddr: spec.MetadataAddr,
	}, nil
})

File: clojure/plugins/voxgig/sekreto/plugins/gcpsecrets.clj
;; GCP Secret Manager, as a voxgig/plugin definition.
;;
;; A plugin rather than a built-in kind because it opens a socket
;; (docs/design/plugin-providers.md).
;;
;; A port of typescript/plugins/gcpsecrets.ts, which is canonical.

(ns voxgig.sekreto.plugins.gcpsecrets
  (:require [voxgig.sekreto.core :as core]
            [voxgig.sekreto.addr :refer [checkaddr]]
            [voxgig.sekreto.json :as json]
            [voxgig.sekreto.provider :as provider]
            [voxgig.sekreto.providers :as providers]
            [voxgig.sekreto.plugins.httpjson :as http])
  (:import [java.nio.charset StandardCharsets]
           [java.util Base64]))

(defrecord Gcpsecrets [project token addr metadataaddr state]
  provider/Provider

  ;; GCP Secret Manager.
  ;;
  ;; `api.token` reads secret `api_token` (dots flattened to `_`; Secret
  ;; Manager ids have no hierarchy and reject dots), latest version. The
  ;; token comes from config, then `GOOGLE_OAUTH_ACCESS_TOKEN`, then the
  ;; GCE/GKE metadata server - so on Google's own platform no credential
  ;; configuration is needed at all.
  ;;
  ;; The metadata call itself is plain http to a link-local host by platform
  ;; design; no credential rides on it, so `checkaddr` guards the Secret
  ;; Manager address instead.
  (lookup [_ name]
    (when (= "" (or project ""))
      (throw (core/sekretoerror "sekreto: gcp: no project")))

    (let [useaddr (http/firstof addr "https://secretmanager.googleapis.com")
          _ (checkaddr useaddr)

          login (fn []
                  (let [configured (http/firstof token (core/getenv "GOOGLE_OAUTH_ACCESS_TOKEN"))]
                    (if (not= "" configured)
                      [configured Long/MAX_VALUE]
                      (let [host (or (core/notempty metadataaddr)
                                     (when-let [named (core/notempty (core/getenv "GCE_METADATA_HOST"))]
                                       (str "http://" named))
                                     "http://metadata.google.internal")
                            url (str (http/trimslash host)
                                     "/computeMetadata/v1/instance/service-accounts/default/token")
                            res (http/fetchjson "GET" url {"Metadata-Flavor" "Google"})
                            got (json/astext (json/dig (:body res) "access_token"))]

                        (when (or (not= 200 (:status res)) (nil? (core/notempty got)))
                          (throw (core/sekretoerror
                                  "sekreto: gcp: no token and metadata server did not answer")))

                        [got (http/renewtime (json/dig (:body res) "expires_in"))]))))

          live (http/withtoken state login)
          url (str (http/trimslash useaddr) "/v1/projects/" project "/secrets/"
                   (core/flatname name "_") "/versions/latest:access")
          res (http/fetchjson "GET" url {"authorization" (str "Bearer " (or live ""))})]

      (cond
        (= 404 (:status res)) nil

        (not= 200 (:status res))
        (throw (core/sekretoerror (str "sekreto: gcp error: " (:status res) ": " url)))

        :else
        (when-let [data (json/asstr (json/dig (:body res) "payload" "data"))]
          (try
            (String. (.decode (Base64/getDecoder) ^String data) StandardCharsets/UTF_8)
            ;; See the aws provider: an undecodable payload is an error.
            (catch IllegalArgumentException _
              (throw (core/sekretoerror "sekreto: gcp: undecodable secret"))))))))

  (describe [_] (str "gcpsecrets:" (or project ""))))

(def gcpsecrets
  "The `gcpsecrets` provider kind."
  (providers/providerplugin
   "gcpsecrets"
   (fn [{:keys [project token addr metadataaddr]}]
     (->Gcpsecrets project token addr metadataaddr
                   (atom {:token nil :renewat Long/MAX_VALUE})))))
